[ASTGen] Add experimental feature to use ASTGen in lieu of parsing types

DougGregor opened the pull request:
Introduce a new experimental feature `ASTGenTypes` that uses ASTGen to translate the Swift syntax tree (produced by the new Swift parser) into C++ `TypeRepr` nodes instead of having the C++ parser create the nodes.

The approach here is to intercept the C++ parser's `parseType` operation to find the Swift syntax node at the given position (where the lexer currently is) and have ASTGen translate that into the corresponding C++ AST node. Then, we spin the lexer forward to the token immediately following the end of the syntax node and continue parsing.

Introduce a collection of small fixes to the mapping of type syntax nodes into `TypeRepr`s:
* Fix crash due to generic arguments having the wrong pointer values (silly enums)
* Support specifiers like `inout`, `consuming`, and `isolated`
* Recognize `Any` and treat it as the empty composition type
* Handle back-ticked identifiers
* Handle variadic parameters in functions and function types
* Handle simple type attributes (this is *not* complete)
* Handle protocol composition types and provide correct source information
* Handle `_` identifier as the empty identifier

With all of these changes, a hacked compiler that enables this experimental flag all the time successfully handles 236 of 245 tests in `test/Sema`.

DougGregor committed:
[ASTGen] Add experimental feature to use ASTGen in lieu of parsing types

Introduce a new experimental feature `ASTGenTypes` that uses ASTGen to
translate the Swift syntax tree (produced by the new Swift parser)
into C++ `TypeRepr` nodes instead of having the C++ parser create the
nodes.

The approach here is to intercept the C++ parser's `parseType`
operation to find the Swift syntax node at the given position (where
the lexer currently is) and have ASTGen translate that into the
corresponding C++ AST node. Then, we spin the lexer forward to the
token immediately following the end of the syntax node and continue
parsing.

DougGregor commented:
@CodaFi you're gonna like this...

DougGregor commented:
@swift-ci please smoke test

CodaFi reviewed (approved):
I do! I do like this!

DougGregor committed:
[ASTGen] Make sure we provide proper TypeReprs for generic arguments

The pointers we were vending to C++ had low bits set because they were
AST node entries rather than the raw TypeReprs.

ahoppen commented on include/swift/Parse/Parser.h:
```diff
@@ -1325,6 +1326,49 @@ class Parser {
+  /// is NULL, something went wrong.
+  template<typename T>
+  using ASTFromSyntaxTreeCallback = std::pair<T*, const void *>(
+      void *sourceFile, const void *sourceLoc
```
We really need to figure out a better way to pass things around than using `void *`.

DougGregor replied:
Yes, this is pretty dreadful. Typedefs would improve on things a little bit. @bnbarham keeps threatening to add them

bnbarham replied:
https://github.com/apple/swift/pull/66044 for the common types (though this would be the first `SourceFile` I think?). Decls/exprs/typerefs/etc are a little more difficult since the visitor just returns a `ASTNode` for each `visit` at the moment. Even disregarding that I'm not sure we want to bridge every single AST node type that we need.

But it would be nice to do something to avoid:
  - https://github.com/apple/swift/pull/66033/commits/682ac98b253e1ef18ffc44e6f6c767a02ca79125 - forgot `.rawValue`
  - https://github.com/apple/swift/pull/66044/files#diff-05684b1c93b20e9444846f61a3575d650a447713a129b477ecf5a518828d17e9L187 - `name` was *not* a `BridgedIdentifier` but rather a `UnresolvedDeclRefExpr`. Finding that involves looking at the visit for `VariableDeclSyntax` -> visits its `pattern` -> visits `IdentifierPatternSyntax` -> calls `SwiftIdentifierExpr_create` -> creates a `UnresolvedDeclRefExpr`.

ahoppen commented on lib/ASTGen/Sources/ASTGen/Macros.swift:
```diff
@@ -635,10 +635,11 @@ func expandFreestandingMacroInProcess(
   sourceLocationPtr: UnsafePointer<UInt8>?,
-  type: Node.Type
+  type: Node.Type,
+  wantOutermost: Bool = false
```
Could you add a comment what, `wantOutermost` does? Also, I think it would also be good to document the other parameters.

DougGregor replied:
Yes, I will. I should really move this API somewhere else (e.g., on `ExportedSourceFile`) because it's in an odd place now that I made it non-`private`.

ahoppen commented on lib/ASTGen/Sources/ASTGen/Macros.swift:
```diff
@@ -664,16 +665,35 @@ private func findSyntaxNodeInSourceFile<Node: SyntaxProtocol>(
+    while let parentSyntax = resultSyntax.parent {
+      guard let typedParent = parentSyntax.as(type),
+            typedParent.position == resultSyntax.position else {
+        break;
```
No semicolon in Swift 😉

---

Also, I’m not entirely sure I understand what `wantOutermost` does. It finds the outermost node that starts at the same location as the innermost node that contains the given position? But then, if you’ve got 
`MyType<MyGenericParam>` and you start looking for the outermost type from `MyGenericParam`, you will get `MyGenericParam` but intuitively I would imagine you to want `MyType<MyGenericParam>`

DougGregor replied:
It's very much focused on the C++ parser interleaving use case, where we want the syntax node that starts at the given location and is a type. The normal behavior of this function finds the innermost syntax node, which is generally why one wants. This one needs to walk further up the tree because if we are at `Swift.Int`, we don't want the `Swift` node, we want the whole `Swift.Int` node. Perhaps there's a better name for this.

ahoppen replied:
But if we start at `Int`, do we want `Swift.Int` or `Int`?

DougGregor replied:
We want `Int`, and that's what we would get because the start position of `Int` and the position of `Swift.Int` are different.

ahoppen commented on lib/ASTGen/Sources/ASTGen/Types.swift:
```diff
@@ -207,3 +207,37 @@ extension ASTGenVisitor {
+  endTypeLocPtr.pointee = sourceFile.pointee.buffer.baseAddress!.advanced(by: typeSyntax.endPosition.utf8Offset)
+
+  // Convert the type syntax node.
+  let typeReprNode = ASTGenVisitor(ctx: ctx, base: sourceFile.pointee.buffer.baseAddress!, declContext: dc)
+    .visit(typeSyntax)
```
Out of curiosity: How complete is ASTGen for types?

DougGregor replied:
A lot of it is there! There is a huge gap where we drop all of the attributes, which breaks a ton of test cases, and I have a few bug fixes locally to pick up specific things that are missing (`inout` doesn't work, there's a bug with generic argument lists that causes a crash, etc.). I'll push a group of those in a bit.

DougGregor replied:
FWIW, I've picked off many of these, but there are still more to do. The biggest remaining grind here is to deal with all of the type attributes that have special rules (`@convention`, `@differentiable`, etc.) and wire up custom attributes.

ahoppen commented on include/swift/Parse/Parser.h:
```diff
@@ -1325,6 +1326,49 @@ class Parser {
+    // Spin the lexer until we get to the ending location.
+    while (Tok.getLoc().getOpaquePointerValue() < endLocPtr &&
+           !Tok.is(tok::eof))
+      consumeToken();
```
I wonder if it’s faster to advance the lexer by as many bytes as the syntax node takes up instead of consuming tokens. IIRC that’s what I did in my previous ASTGen implementation.

DougGregor replied:
Hmm. Advance the lexer and then prime it again; that's better than what I'm doing. Do you have the code lying around?

ahoppen replied:
This is the code I used for incremental SwiftSyntax parsing

https://github.com/apple/swift/blob/f3fd46270289b62fa10802f6b5925cbf22290232/lib/Parse/Parser.cpp#L837-L855

DougGregor committed:
[ASTGen] Support specifiers on types such as `inout` and `consuming`

DougGregor committed:
[ASTGen] Handle the 'Any' type

DougGregor committed:
[ASTGen] Handle back-ticked names as identifiers

DougGregor committed:
[ASTGen] Handle variadic parameters in functions and function types

There is a modeling difference between the swift-syntax tree and the
C++ type representation (TypeRepr) that is a little odd here, so we
end up parsing the ellipsis on the C++ side rather than looking "up"
the syntax tree to find it.

DougGregor committed:
[ASTGen] Handle simple type attributes.

Bridge the simple type attributes like `@autoclosure` into an
`AttributedTypeRepr`. We don't validate the arguments, and we don't
handle more complicated attributes like `@convention(c)` just yet.

DougGregor committed:
[ASTGen] Handle composition types parsed through `parseType()`.

Protocol composition types such as `P & Q` are similar to nested types
like `A.B` because the innermost type syntax node at the given position
doesn't cover the whole type. Adjust by looking back up the tree.
This all feels like a hack, and there should be a better way.

While here, fix the source ranges passed in to create
`CompositionTypeRepr`. We were tripping assertions due to missing
source-location information.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[ASTGEn] Handle `_` identifier as the empty identifier

DougGregor committed:
[Parser] Reset C++ lexer just past the whole ASTGen-parsed syntax node

Instead of "spinning" the C++ lexer, consuming tokens uptil we get past
the point where ASTGen told us the end of the syntax node was, just
reset the lexer to exactly that position. This is more efficient, but
also fixes problems where we would end up skipping past a `>` that had
been split.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
Update to use the exported-source-file request

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
I'm going to merge this and we'll iterate to get it into a nicer form before writing more parsing logic
